nemo/collections/asr/parts/submodules/adapters/attention_adapter_mixin.py

```diff
@@ -0,0 +1,119 @@
+import torch
+
+from nemo.core.classes.mixins import adapter_mixins
+from nemo.utils import logging, logging_mode
+
+
+class AttentionAdapterModuleMixin(adapter_mixins.AdapterModuleMixin):
+    """
+    Utility class that implements a custom forward method for Modules that are attention based.
+    Attention based adapters can support either linear adapters, and Multi-Head Attention adapters.
+
+    However, Multi Head Attention adapters require additional arguments, such as `att_mask` and `pos_emb`.
+    This utility class unifies the adapter forward pass for both types of adapters.
+
+    .. Usage:
+
+        To use this class, inherit from this class, and when calling self.foward_enabled_adapters() pass the following:
+
+    .. code-block:: python
+
+            if self.is_adapter_available():
+                # Call the MHA adapters
+                pack_ip = {
+                    'x': residual,
+                    'loc': 'mha',
+                    'att_mask': att_mask,
+                    'pos_emb': pos_emb,
+                }
+                pack_ip = self.forward_enabled_adapters(pack_ip)
+                residual = pack_ip['x']
+
+            if self.is_adapter_available():
+                # Call the Linear adapters
+                pack_ip = {
+                    'x': x,
+                    'loc': 'post',
+                }
+                pack_ip = self.forward_enabled_adapters(pack_ip)
+                x = pack_ip['x']
+    """
+
+    def forward_single_enabled_adapter_(
+        self,
+        input: dict,
+        adapter_module: torch.nn.Module,
+        *,
+        adapter_name: str,
+        adapter_strategy: 'nemo.core.classes.mixins.adapter_mixin_strategies.AbstractAdapterStrategy',
+    ):
+        """
+        Perform the forward step of a single adapter module on some input data.
+
+        **Note**: Subclasses can override this method to accommodate more complicate adapter forward steps.
+
+        Args:
+            input: Dictionary of packed tensors. The dict should contain at least
+                `x`: output tensor
+                `loc`: Semantic location in module where this adapter was called
```

Comment:
can you add current possible values for `loc`

Reply:
Good catch